### PR TITLE
fix e2e service invocation app to return status codes for non 2xx errors

### PR DESCRIPTION
**Pull request by vinayada1**
# Description

1. Change service invocation e2e test app to look at status code and not just err
2. Add http to grpc test for allowlists

#2158 #2121 

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2158 #2121 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: tests/e2e/allowlists/allowlists_service_invocation_test.go [hunk 1]
```diff
@@ -1,5 +1,3 @@
-// +build e2e
```
Comment (mukundansundar) [inline tests/e2e/allowlists/allowlists_service_invocation_test.go:1]: add back build e2e.

**Review by mukundansundar — CHANGES_REQUESTED** [review]

File: tests/e2e/allowlists/allowlists_service_invocation_test.go (hunk 1 above)
Comment (vinayada1) [reply tests/e2e/allowlists/allowlists_service_invocation_test.go]: Thanks

File: tests/apps/service_invocation/app.go [hunk 2]
```diff
@@ -1066,6 +1104,24 @@ func HTTPPost(url string, data []byte) ([]byte, error) {
 	return extractBody(resp.Body)
 }
 
+// HTTPPostWithStatus is a helper to make POST request call to url
+func HTTPPostWithStatus(url string, data []byte) ([]byte, int, error) {
```
Comment (mukundansundar) [inline tests/apps/service_invocation/app.go:1108]: seems an unused method.

File: tests/apps/service_invocation/app.go [hunk 3]
```diff
@@ -913,6 +922,35 @@ func httpWrapper(httpMethod string, url string, data []byte) (appResponse, error
 	return appResp, nil
 }
 
+func httpWrapperWithStatus(httpMethod string, url string, data []byte) (appResponse, error) {
```
Comment (mukundansundar) [inline tests/apps/service_invocation/app.go:925]: seems an unused method.

File: tests/e2e/allowlists/allowlists_service_invocation_test.go [hunk 4]
```diff
@@ -88,25 +89,26 @@ var allowlistsServiceinvocationTests = []struct {
...
-	// 	"Test deny with callee side grpc",
-	// 	"allowlists-callee-grpc",
-	// 	"httptogrpctest",
-	// 	"rpc error: code = PermissionDenied desc = access control policy has denied access to appid: allowlists-caller operation: httptogrpctest verb: NONE",
-	// },
+	{
+		"Test deny with callee side grpc",
+		"allowlists-callee-grpc",
+		"httptogrpctest",
+		http.StatusInternalServerError,
```
Comment (mukundansundar) [inline tests/e2e/allowlists/allowlists_service_invocation_test.go:109]: shouldn't this be 403?

Comment (youngbupark) [reply tests/e2e/allowlists/allowlists_service_invocation_test.go:109]: I think this is test app response code. we need to ensure that we bypass status code. 

When I double-check the conversion, PermissionDenied should be converted to 403.
https://github.com/dapr/dapr/blob/67c4553aea13acef634540469f4a287b759a288e/pkg/messaging/v1/util.go#L243-L244

File: tests/e2e/allowlists/allowlists_service_invocation_test.go [hunk 5]
```diff
@@ -88,25 +89,26 @@ var allowlistsServiceinvocationTests = []struct {
 }
 
 var moreAllowlistsServiceinvocationTests = []struct {
```
Comment (youngbupark) [inline tests/e2e/allowlists/allowlists_service_invocation_test.go:91]: Can you rename `moreAllowlistsServiceinvocationTests` to be more specific ?

I would refactor `allowlistsServiceinvocationTests` and combine this to `allowlistsServiceinvocationTests`. 

I can see the duplicated test code too. please remove duplicated codes and combine it to single one.

**Review by youngbupark — CHANGES_REQUESTED** [review]

File: tests/e2e/allowlists/allowlists_service_invocation_test.go (hunk 4 above)
Comment (vinayada1) [reply tests/e2e/allowlists/allowlists_service_invocation_test.go:109]: Yes, the test app is doing this

File: tests/e2e/allowlists/allowlists_service_invocation_test.go (hunk 5 above)
Comment (vinayada1) [reply tests/e2e/allowlists/allowlists_service_invocation_test.go:91]: fixed

File: pkg/http/api.go [hunk 6]
```diff
@@ -591,8 +592,14 @@ func (a *api) onDirectMessage(reqCtx *fasthttp.RequestCtx) {
 	resp, err := a.directMessaging.Invoke(reqCtx, targetID, req)
 	// err does not represent user application response
 	if err != nil {
+		// Allowlists policies that are applied on the callee side can return a Permission Denied error.
+		// For everything else, treat it as a gRPC transport error
+		statusCode := fasthttp.StatusInternalServerError
+		if status.Code(err) == codes.PermissionDenied {
+			statusCode = invokev1.HTTPStatusFromCode(codes.PermissionDenied)
```
Comment (mukundansundar) [inline pkg/http/api.go:599]: @wcs1only This is a change that relates to https://github.com/dapr/dapr/issues/2338

**Comment by mukundansundar** [comment]
@vinayada1 Please add unit test to the changes that you made to `http/api.go`

**Comment by vinayada1** [comment]
> @vinayada1 Please add unit test to the changes that you made to `http/api.go`

Done

File: pkg/http/api.go (hunk 6 above)
Comment (wcs1only) [reply pkg/http/api.go:599]: Noted. If you merge before me, I'll be sure be sure to take a look.

Comment (vinayada1) [reply pkg/http/api.go:599]: Thanks!

File: pkg/http/api_test.go [hunk 7]
```diff
@@ -469,6 +471,36 @@ func TestV1DirectMessagingEndpoints(t *testing.T) {
 		assert.Equal(t, "ERR_DIRECT_INVOKE", resp.ErrorBody["errorCode"])
 	})
 	fakeServer.Shutdown()
```
Comment (mukundansundar) [inline pkg/http/api_test.go:473]: why is fakeServer shut down here?

Comment (vinayada1) [reply pkg/http/api_test.go]: removed

**Review by mukundansundar — APPROVED** [review]

**Review by youngbupark — APPROVED** [review]
lgtm

**Pull request merged**